### ISSUE TITLE
FileBase, DirectoryBase, PathBase, etc. should expose an IFileSystem property

### ISSUE THREAD

**Issue by herebebeasties** [System-IO-Abstractions/System.IO.Abstractions#351]
We have an internal extensions library for `IFileSystem` that adds various nice extension methods like `Unzip(...)` and `DeleteDirectoryContents(...)`.

Because these need to use operations on a mixture of `FileBase`, `DirectoryBase` and `PathBase`, these currently need to be extension methods on `IFileSystem` which isn't ideal:
 * It's much less discoverable, particularly for things that really ought to live alongside a closely-related API like `DirectoryBase.Delete`.
 * It causes API breaks if you add functionality to an extension method on `FileBase` and suddenly discover you need to move it to an `IFileSystem` one instead.

It would be extremely useful to expose the underlying `IFileSystem` on all these base classes as a property, which would fix this.

It would be a pretty trivial change, obviously, although as with any API addition would break anyone who has a custom implementations of FileBase, etc. as obviously they'd need to add it in (pretty trivial to fix, though).

Thoughts?

**Comment by fgreinacher** [comment]
IMO that’s a good idea which is very similar to what’s proposed in #239.

**Comment by herebebeasties** [comment]
OK, great - am happy to work up a PR. It looks like adding this to DirectoryInfo / FileInfo stuff is a bit tricky, because they have an implicit casting operator:
```c#
public static implicit operator DirectoryInfoBase(DirectoryInfo directoryInfo)
public static implicit operator FileInfoBase(FileInfo fileInfo)
```
On the basis that this implicitly casts things to the FileSystem implementation, which may well not be what you want anyway, I'm tempted to just new up a `new FileSystem()` for those. (I don't think having implicit casts like this is a good idea anyway, as it'll silently do the wrong thing if you actually meant for all your usages to use something like a `LongPathFileSystem`.)

Also - `DirectoryWrapper`, `FileWrapper`, etc. are marked as `[Serializable]`. If they suddenly have an instance field for the `IFileSystem` then that will be serialized as well. I can see this was added in #22 by @ericnewton76 but it doesn't have much context as to why. Can anyone shed any light on why you'd want to serialize a `DirectoryBase` instance? It doesn't make much sense to me - feels similar to trying to serialize a Socket or a FileStream or a DNS resolver, or something else ephemeral like that.

**Comment by fgreinacher** [comment]
Implemented by #352
